cmd/query/app/server.go

```diff
@@ -160,7 +161,8 @@ func createHTTPServer(querySvc *querysvc.QueryService, metricsQuerySvc querysvc.
 	}
 
 	ctx, closeGRPCGateway := context.WithCancel(context.Background())
-	if err := apiv3.RegisterGRPCGateway(ctx, logger, r, queryOpts.BasePath, queryOpts.GRPCHostPort, queryOpts.TLSGRPC); err != nil {
+	tc := tenancy.NewTenancyConfig(&queryOpts.Tenancy)
```

Comment:
what about the API handler above, I assume it should also be wrapped in ExtractTenant

Reply:
Thanks for your review.  *This comment was extremely helpful and identified several flaws.*

I have fixed the problems, but please don't merge yet! ;-)

The problems were not detected for two reasons.  I was testing with GetTrace, a streaming RPC, and ignoring tenancy for GetService, a unary RPC.  This PR needs unary RPC tests.  I have added one but need to increase coverage.
The other problem is that the tests set up the server using `newGRPCServer()` in cmd/query/app/grpc_handler_test.go, but the actual Jaeger uses `createGRPCServer()` which is nearly the same in implementation but was not tested for tenancy and didn't work.

I want to do at least one more commit of additional tests.

cc @pavolloffay

Reply:
I am not familiar with `newGRPCServer` nor `createGRPCServer` but if there is a way to unify these APIs we should consider doing it (e.g. in a follow up PR).

Reply:
> `tc := tenancy.NewTenancyConfig(&queryOpts.Tenancy)`

there are way too many places doing this conversion - we should only create tenancy config once (in the future it could be "live" object that self-updates).

Reply:
@yurishkuro. I want to get this right.

I was using TenancyOptions for something human-readable, for Viper, and TenancyConfig for something machine readable, optimized for fast checking, following the pattern of `tlscfg.Options` and `tls.Config`.  Would it make sense to change tenency.TenancyConfig to `tenancy.TenancyManager`?

Currently the PR converts tenancy.Options to TenancyConfig in 4  non-test places.  Twice in `createGRPCServer()` -- I will fix that, once in `createHTTPServer()`, and once in `BuildHandlers()` (already merged; for trace ingestion).  Currently TenancyConfig is lightweight, and I didn't want to change the signatures to the create*Server() functions.  If we think of Tenancy is a "service" then I should create it once.  The next commit will address this.

Reply:
`TenancyConfig` is different from `tls.Config` in that the former is an object with behavior (an internal "service"), but the latter is just data passed elsewhere. I think from the naming it would be clearer if it was called `tenancy.TenancyManager`.

> Currently TenancyConfig is lightweight

It's only lightweight in the current implementation. Because it's a service, it can become arbitrarily complicated in the future, e.g. it could start refreshing the list of tenants from some storage, etc. By instantiating it in many places we introduce technical debt from the start that will need to be cleaned up later, so I'd rather fix it now and create this service once.